### PR TITLE
Add custom MessagePack resolver for mod settings dictionary

### LINKED ISSUE #11660: Changing a mod customisation value causes errors at the beginning of gameplay
**Describe the bug:** If you choose a mod, modifying at least one of its customization values, the game returns various exceptions.

**Screenshots or videos showing encountered issue:** 
The example shows in order: no mods, difficulty adjust mod (default values), difficulty adjust mod (with 1 modified customisation)

https://user-images.githubusercontent.com/43073074/106580257-f903fd80-6541-11eb-8e54-de799f127b18.mp4

**osu!lazer version:** 2021.129.0

**Logs:**
[runtime.log](https://github.com/ppy/osu/files/5910058/runtime.log)

### PR DESCRIPTION
**Pull request by peppy**
This was failing until now (causing potential runtime crashes) when attempting to play with a mod that had adjusted settings (due to MessagePack now knowing how to serialises bindables).

Note that this is not in a working state. While the tests pass, it causes disconnections/errors at a server-side level which I am still investigating. I'm PRing this because I spent the good part of an hour investigating and I'm 99.99% sure there are no issues with my code.

```
dbug: Microsoft.AspNetCore.SignalR.HubConnectionHandler[2]
      Error when processing requests.
      System.IO.InvalidDataException: Reading array length for 'streamIds' failed.
       ---> System.InvalidOperationException: code is invalid. code:202 format:float 32
         at MessagePack.Decoders.InvalidArrayHeader.Read(Byte[] bytes, Int32 offset, Int32& readSize)
         at Microsoft.AspNetCore.SignalR.Protocol.MessagePackHubProtocol.ReadArrayLength(Byte[] input, Int32& offset, String field)
         --- End of inner exception stack trace ---
         at Microsoft.AspNetCore.SignalR.Protocol.MessagePackHubProtocol.ReadArrayLength(Byte[] input, Int32& offset, String field)
         at Microsoft.AspNetCore.SignalR.Protocol.MessagePackHubProtocol.ReadStreamIds(Byte[] input, Int32& offset)
         at Microsoft.AspNetCore.SignalR.Protocol.MessagePackHubProtocol.CreateInvocationMessage(Byte[] input, Int32& offset, IInvocationBinder binder, IFormatterResolver resolver, Int32 itemCount)
         at Microsoft.AspNetCore.SignalR.Protocol.MessagePackHubProtocol.ParseMessage(Byte[] input, Int32 startOffset, IInvocationBinder binder, IFormatterResolver resolver)
         at Microsoft.AspNetCore.SignalR.Protocol.MessagePackHubProtocol.TryParseMessage(ReadOnlySequence`1& input, IInvocationBinder binder, HubMessage& message)
         at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.DispatchMessagesAsync(HubConnectionContext connection)
         at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.RunHubAsync(HubConnectionContext connection)
2021-02-03 10:46:19 [3504]: Disconnected
```

The exception points at the offset not being incremented correctly for the next read (`streamIds` is part of the parent object in signalr's messagepack message protocol), but I have confirmed that the returned `readSize` matches on serialisation and deserialisation of the same message, both at a client-client level and also client-server.

My next step will be to test after updating SignalR to see if it is actually a SignalR (or [Microsoft.AspNetCore.SignalR.Protocol.MessagePackHubProtocol](https://github.com/dotnet/aspnetcore/blob/c3acdcac86dad91c3d3fbc3b93ecc6b7ba494bdc/src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs#L679)) bug.

- Closes #11660

### CODE REVIEW

File: osu.Game/Online/API/ModSettingsDictionaryFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+            var output = new Dictionary<string, object>();
+
+            int itemCount = MessagePackBinary.ReadArrayHeader(bytes, offset, out readSize);
+            offset += readSize;
+
+            for (int i = 0; i < itemCount; i++)
+            {
+                var key = MessagePackBinary.ReadString(bytes, offset, out readSize);
+                offset += readSize;
```
Comment (peppy) [inline osu.Game/Online/API/ModSettingsDictionaryFormatter.cs:55]: Note that this is a pretty shoddy API. It's improved in the newer version of messagepack. I've implemented this copying the way the official formatters are implemented in the previous version, as there is no direct documentation on the interface itself.

You can find examples here: https://github.com/neuecc/MessagePack-CSharp/blob/v1.7.3.7/src/MessagePack/Formatters/PrimitiveFormatter.cs / https://github.com/neuecc/MessagePack-CSharp/blob/v1.7.3.7/src/MessagePack.UnityShims/Formatters.cs

File: osu.Game/Online/API/ModSettingsDictionaryFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+                switch (MessagePackBinary.GetMessagePackType(bytes, offset))
```
Comment (peppy) [inline osu.Game/Online/API/ModSettingsDictionaryFormatter.cs:57]: During my investigation I found https://github.com/neuecc/MessagePack-CSharp/blob/v1.7.3.7/src/MessagePack/Formatters/PrimitiveObjectFormatter.cs, which can likely replace this code. I will update it to use this once I figure out what the cause of the remaining issue is.

**Comment by peppy** [comment]
For reference, here's the 5.0 API, which is about 5 magnitudes nicer

https://github.com/ppy/osu/compare/master...peppy:fix-mod-settings-serlisation-signalr50?expand=1

**Comment by bdach** [comment]
API is indeed pretty bare-bones, but the implementation looks all good as far as I can see. The index/offset math (worst part of all of this, probably) seems to check out.

**Review by bdach — DISMISSED** [review]

**Review by frenzibyte — CHANGES_REQUESTED** [review]
`ModEasyWithExtraLives` is still suffering from this issue, as it defines a setting of type `BindableInt` which is not handled in the formatter here.

It all feels like bindable should expose an `object` value somehow, but given that it's been rejected multiple times for reasons I can't recall, going with switch cases is the way.

**Comment by peppy** [comment]
I'm not going to update this PR because it still doesn't work. Will be testing against iOS today and PR the 50 version (https://github.com/ppy/osu/compare/master...peppy:fix-mod-settings-serlisation-signalr50?expand=1) if I can get everything working.

**Comment by peppy** [comment]
Fixes have been applied on #11673.

Closing this as it doesn't work due to upstream bugs.

**Pull request closed without merging**